File: ddingmin/baekjoon/14925.py
import sys


def solve(n, m, arr):
    dp = [[0] * m for _ in range(n)]
    ans = 0

    for i in range(n):
        for j in range(m):
            if arr[i][j] == 0:
                dp[i][j] = min(dp[i - 1][j - 1], dp[i - 1][j], dp[i][j - 1]) + 1
                ans = max(ans, dp[i][j])

    return ans


def main():
    n, m = map(int, input().split())
    arr = [list(map(int, input().split())) for _ in range(n)]
    print(solve(n, m, arr))


if __name__ == '__main__':
    input = sys.stdin.readline
    main()

File: ddingmin/baekjoon/14925.kt
import java.io.BufferedWriter
import java.io.OutputStreamWriter

fun main() = with(System.`in`.bufferedReader()) {
    val sout = BufferedWriter(OutputStreamWriter(System.out))

    var nm = readLine().split(" ").map { it.toInt() }

    var arr = Array(nm[0]) {
        readLine().split(" ").map { it.toInt() }
    }

    sout.appendLine(solve(nm, arr))
    sout.close()
}

fun solve(nm: List<Int>, arr: Array<List<Int>>): String {
    var ans = 0
    var dp = Array(nm[0] + 1) {
        IntArray(nm[1] + 1)
    }

    for (i in 1..nm[0]) {
        for (j in 1..nm[1]) {
            if (arr[i - 1][j - 1] == 0) dp[i][j] = minOf(dp[i - 1][j - 1], dp[i - 1][j], dp[i][j - 1]) + 1
            ans = maxOf(ans, dp[i][j])
        }
    }

    return ans.toString()
}
